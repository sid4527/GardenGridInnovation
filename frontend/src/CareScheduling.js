import React, { useState } from 'react'; 
import { useNavigate, Link } from 'react-router-dom';
import './CareScheduling.css';

function CareScheduling() {
  const [showForm, setShowForm] = useState(false);
  const [formData, setFormData] = useState({
    date: '',
    plant: 'Rose Bush',
    careType: 'Watering',
    notes: '',
  });
  const [tasks, setTasks] = useState([ // State to store tasks
    { date: '2024-11-01', plant: 'Rose Bush', careType: 'Watering', status: 'Completed' },
    { date: '2024-11-02', plant: 'Lavender', careType: 'Fertilizing', status: 'Pending' },
  ]);
  const navigate = useNavigate();

  const handleNewEntry = () => {
    setShowForm(!showForm);
  };

  const handleLoginRedirect = () => {
    navigate('/');
  };

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData({ ...formData, [name]: value });
  };

  const handleSubmit = async (e) => {
    e.preventDefault();

    const newTask = {
      date: formData.date,
      plant: formData.plant,
      careType: formData.careType,
      status: 'Pending', // Default status on submission
    };

    try {
      const response = await fetch('http://localhost:9000/api/care-tasks', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          TaskDate: formData.date,
          PlantName: formData.plant,
          CareType: formData.careType,
          Status: 'Pending',
          Notes: formData.notes,
        }),
      });

      if (response.ok) {
        console.log('New care task added successfully');
        setTasks([...tasks, newTask]); // Append the new task to the list
        setShowForm(false); // Hide the form after successful submission
        setFormData({ date: '', plant: 'Rose Bush', careType: 'Watering', notes: '' }); // Reset the form
      } else {
        console.error('Failed to add new care task');
      }
    } catch (error) {
      console.error('Error:', error);
    }
  };

  return (
    <div className="care-scheduling">
      <header className="navbar">
        <div className="navbar-content">
          <div className="navbar-logo">ðŸŒ¿ Garden Grid</div>
          <nav className="navbar-links">
            <Link to="/">Home</Link>
            <Link to="/inventory">Inventory</Link>
            <Link to="/inventory-management">Entries</Link>
            <Link to="/resources">Resources</Link>
            <Link to="/reports">Reports</Link>
            <Link to="/login">Log Out</Link>
          </nav>
        </div>
      </header>

      <nav className="tabs">
        <Link to="/inventory-management" className="tab">Inventory Management</Link>
        <Link to="/care-scheduling" className="tab">Care Scheduling</Link>
        <Link to="/growth-tracking" className="tab">Growth Tracking</Link>
      </nav>

      <div className="content">
        <h2>Care Scheduling</h2>
        <button className="new-entry-button" onClick={handleNewEntry}>+ New Entry</button>

        <table className="scheduling-table">
          <thead>
            <tr>
              <th>Date</th>
              <th>Plant</th>
              <th>Care Type</th>
              <th>Status</th>
            </tr>
          </thead>
          <tbody>
            {tasks.map((task, index) => (
              <tr key={index}>
                <td>{task.date}</td>
                <td>{task.plant}</td>
                <td>{task.careType}</td>
                <td>{task.status}</td>
              </tr>
            ))}
          </tbody>
        </table>

        {showForm && (
          <div className="entry-form">
            <h3>Add New Care Task</h3>
            <form onSubmit={handleSubmit}>
              <label>
                Date: <input type="date" name="date" value={formData.date} onChange={handleChange} required />
              </label>
              <label>
                Plant:
                <select name="plant" value={formData.plant} onChange={handleChange}>
                  <option value="Rose Bush">Rose Bush</option>
                  <option value="Lavender">Lavender</option>
                </select>
              </label>
              <label>
                Care Type:
                <select name="careType" value={formData.careType} onChange={handleChange}>
                  <option value="Watering">Watering</option>
                  <option value="Fertilizing">Fertilizing</option>
                  <option value="Pruning">Pruning</option>
                </select>
              </label>
              <label>
                Notes: <input type="text" name="notes" value={formData.notes} onChange={handleChange} placeholder="Any special notes..." />
              </label>
              <button type="submit" className="submit-button">Submit</button>
            </form>
          </div>
        )}

      </div>
    </div>
  );
}

export default CareScheduling;
