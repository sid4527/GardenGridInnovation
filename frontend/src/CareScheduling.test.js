import React from 'react';
import { render, screen, fireEvent, waitFor, act } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import axios from 'axios';
import CareScheduling from './CareScheduling';

// Mock axios for API calls
jest.mock('axios', () => ({
  post: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

describe('CareScheduling Component', () => {
  test('renders header with correct title', () => {
    act(() => {
      render(
        <MemoryRouter>
          <CareScheduling />
        </MemoryRouter>
      );
    });
    expect(screen.getByText('Garden Grid Care Scheduling')).toBeInTheDocument();
  });

  test('renders navigation tabs', () => {
    render(
      <MemoryRouter>
        <CareScheduling />
      </MemoryRouter>
    );

    const navTabs = screen.getAllByRole('link');
    expect(navTabs).toHaveLength(3);
    expect(navTabs[1]).toHaveTextContent('Care Scheduling');
    expect(screen.getByRole('heading', { name: 'Care Scheduling' })).toBeInTheDocument();
  });

  test('toggles new entry form visibility', () => {
    render(
      <MemoryRouter>
        <CareScheduling />
      </MemoryRouter>
    );

    const newEntryButton = screen.getByText('+ New Entry');
    fireEvent.click(newEntryButton);
    expect(screen.getByText('Add New Care Task')).toBeInTheDocument();

    fireEvent.click(newEntryButton);
    expect(screen.queryByText('Add New Care Task')).not.toBeInTheDocument();
  });

  test('fills and submits the new entry form', async () => {
    axios.post.mockResolvedValueOnce({ data: { success: true } });

    render(
      <MemoryRouter>
        <CareScheduling />
      </MemoryRouter>
    );

    fireEvent.click(screen.getByText('+ New Entry'));

    fireEvent.change(screen.getByLabelText('Date:'), { target: { value: '2024-11-20' } });
    fireEvent.change(screen.getByLabelText('Plant:'), { target: { value: 'Lavender' } });
    fireEvent.change(screen.getByLabelText('Care Type:'), { target: { value: 'Pruning' } });
    fireEvent.change(screen.getByLabelText('Notes:'), { target: { value: 'Trim before frost' } });

    fireEvent.click(screen.getByText('Submit'));

    await waitFor(() => {
      expect(axios.post).toHaveBeenCalledWith('/api/care-scheduling', {
        date: '2024-11-20',
        plant: 'Lavender',
        careType: 'Pruning',
        notes: 'Trim before frost',
      });
    });
  });

  test('shows error message for invalid form submission', async () => {
    render(
      <MemoryRouter>
        <CareScheduling />
      </MemoryRouter>
    );

    fireEvent.click(screen.getByText('+ New Entry'));
    fireEvent.click(screen.getByText('Submit'));

    await waitFor(() => {
      expect(screen.getByText(/All fields are required/i)).toBeInTheDocument();
    });
  });

  test('navigates back to home', () => {
    const mockedNavigate = jest.fn();
    jest.mock('react-router-dom', () => ({
      ...jest.requireActual('react-router-dom'),
      useNavigate: () => mockedNavigate,
    }));

    render(
      <MemoryRouter>
        <CareScheduling />
      </MemoryRouter>
    );

    fireEvent.click(screen.getByText('Back to Home'));
    expect(mockedNavigate).toHaveBeenCalledWith('/');
  });

  test('matches snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <CareScheduling />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
