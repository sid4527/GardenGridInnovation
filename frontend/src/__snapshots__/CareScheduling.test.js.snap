// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CareScheduling Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="care-scheduling"
    role="main"
  >
    <header
      class="header"
    >
      <h1>
        Garden Grid Care Scheduling
      </h1>
    </header>
    <nav
      aria-label="Navigation Tabs"
      class="tabs"
    >
      <a
        class="tab"
        href="/inventory-management"
      >
        Inventory Management
      </a>
      <a
        class="tab"
        href="/care-scheduling"
      >
        Care Scheduling
      </a>
      <a
        class="tab"
        href="/growth-tracking"
      >
        Growth Tracking
      </a>
    </nav>
    <div
      class="content"
    >
      <h2>
        Care Scheduling
      </h2>
      <button
        aria-expanded="false"
        class="new-entry-button"
      >
        + New Entry
      </button>
      <table
        class="scheduling-table"
        role="table"
      >
        <thead>
          <tr>
            <th
              scope="col"
            >
              Date
            </th>
            <th
              scope="col"
            >
              Plant
            </th>
            <th
              scope="col"
            >
              Care Type
            </th>
            <th
              scope="col"
            >
              Status
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              2024-11-01
            </td>
            <td>
              Rose Bush
            </td>
            <td>
              Watering
            </td>
            <td>
              Completed
            </td>
          </tr>
          <tr>
            <td>
              2024-11-02
            </td>
            <td>
              Lavender
            </td>
            <td>
              Fertilizing
            </td>
            <td>
              Pending
            </td>
          </tr>
        </tbody>
      </table>
      <button
        class="login-button"
      >
        Back to Home
      </button>
    </div>
  </div>
</DocumentFragment>
`;
